components/navigation/PreventNavigation.vue

```diff
@@ -0,0 +1,124 @@
+    >
+      <template #header>
+        <h5 class="modal-title">
+          <slot name="header">
+            Warning
+          </slot>
+        </h5>
+      </template>
+
+      <template>
+        <slot />
+      </template>
+
+      <template #footer>
+        <button
+          v-for="action in actions"
+          :key="action.text"
+          :ref="`actionBtn-${action.text}`"
+          type="button"
+          :class="`btn btn-${action.type || 'secondary'}`"
+          @click="actionClick(action)"
+        >
+          {{ action.text }}
+        </button>
+      </template>
+    </CModal>
+  </div>
+</template>
+<script>
+
+import { CModal } from '@coreui/vue';
+
+export default {
+  name      : 'PreventNavigation',
+  components: { CModal },
+  props     : {
+    prevent: {
+      type   : Boolean,
+      default: false,
+    },
+    actions: {
+      type    : Array,
+      required: true,
+      validator(arr) {
+        const regex = /[a-z]+/;
+        return arr.length > 0
+          && arr.every((action) => action.text && action.name
+          && regex.test(String(action.name)));
+      },
+    },
+  },
+  data() {
+    return {
+      isOpen     : false,
+      routeParams: {
+        to  : null,
+        from: null,
+        next: null,
+      },
+      unregisterMyRouterGuardListener: () => {},
+    };
+  },
+  mounted() {
+    // register system unload event (eg: if user  type a new URL)
+    window.addEventListener('beforeunload', this.preventNav);
+    // register vuejs router guard listener
+    const unregister = this.$router.beforeEach((to, from, next) => { // must be in an arrow function to save `this` closure reference
+      beforeRouteChange.call(this, { to, from, next });
+    });
+
+    this.unregisterMyRouterGuardListener = unregister; // save unregister fn for beforeDestroy
+  },
+
+  beforeDestroy() {
+    // unregister system unload event handler
+    window.removeEventListener('beforeunload', this.preventNav);
+
+    // unregister my router guard listerner
+    this.unregisterMyRouterGuardListener();
+  },
+
+  methods: {
+    actionClick,
+    setFocus,
+    preventNav(e) {
+      if (!this.prevent) return;
+      const event = e;
+      event.preventDefault();
+      event.returnValue = '';
+    },
+  },
+};
+
+function setFocus() {
+  const selectedAction = this.actions.find((action) => action.default && action.default === true)
+  || this.actions.find((action) => action.type && action.type === 'primary')
+  || this.actions.find((action) => action.type && action.type === 'success')
+  || this.actions[0];
+
+  this.$nextTick(() => {
+    this.$refs[`actionBtn-${selectedAction.text}`][0].focus();
+  });
+}
+
+async function beforeRouteChange(routeParams) {
+  if (this.prevent) {
+    this.routeParams = routeParams;
+    this.isOpen      = true;
+    this.setFocus();
+  } else {
+    routeParams.next();
+  }
+}
+
+function actionClick({ name }) {
+  this.isOpen = false;
+  this.$emit(name, this.routeParams);
```

Comment:
Better to clone the object reference to avoid parent/child manipulation
```suggestion
  const params = {...this.routeParams }
  this.$emit(name, params);
```

Reply:
DONE! changes done as suggested.